### ISSUE TITLE
[FEATURE REQUEST] - Variable hoisting in decision trees.

### ISSUE THREAD

**Issue by jordwalke** [ocsigen/js_of_ocaml#945]
`generate.ml` currently compiles decision trees like this:

```javascript
function extension_constructor(x) {
  if (is_block(x))
    if (caml_obj_tag(x) !== 248)
      if (1 <= x.length - 1) {
        var slot = x[1];
        var switch__0 = 1;
      } else {
        var switch__0 = 0;
      }
    } else {
      var switch__0 = 0;
    }
  } else {
    var switch__0 = 0;
  }
  if (!switch__0) {
    var slot = x;
  }
  if (is_block(slot))
    if (caml_obj_tag(slot) === 248) {
      var name = slot[1];
      var switch__1 = 1;
    } else {
      var switch__1 = 0;
    }
  } else {
    var switch__1 = 0;
  }
  if (!switch__1) {
    var name = call1(Pervasives[1], cst_Obj_extension_constructor__0);
  }
  return caml_obj_tag(name) === 252
    ? slot
    : call1(Pervasives[1], cst_Obj_extension_constructor);
}
```

This is limiting and problematic because the variables are declared deep in the branches, and not every branch declares the same set of variables. jsoo relies on the fact that (in this case) "if `switch__0` was not set to `1`, then we know that `slot` was not set and so we'll set it after the loop". This has been troublesome for us because various language linters and type checkers will get tripped up on the fact that it looks like `slot` may not always be declared. Languages like PHP and JS actually have function scope, so these variables are declared at the top of the function as undefined whether or not they were explicitly declared - their mere presence in the body of the function make it equivalent to `var slot = undefined`. For extending jsoo to different backends, it would make it much easier if the variables were hoisted up to the top of the function, even if they had `None` expression at the time of `Variable_declaration` at the top of the function.

Instead, I could imagine:

```javascript
function extension_constructor(x) {
  ...
  // Hoisted as high as necessary.
  var slot, switch__0, switch__1, name;
  if (is_block(x))
    if (caml_obj_tag(x) !== 248)
      if (1 <= x.length - 1) {
        slot = x[1];
        switch__0 = 1;
      } else {
        switch__0 = 0;
      }
    } else {
      switch__0 = 0;
    }
  } else {
    switch__0 = 0;
  }
  if (!switch__0) {
    slot = x;
  }
  if (is_block(slot))
    if (caml_obj_tag(slot) === 248) {
      name = slot[1];
      switch__1 = 1;
    } else {
      switch__1 = 0;
    }
  } else {
    switch__1 = 0;
  }
  if (!switch__1) {
    name = call1(Pervasives[1], cst_Obj_extension_constructor__0);
  }
  return caml_obj_tag(name) === 252
    ? slot
    : call1(Pervasives[1], cst_Obj_extension_constructor);
}
```

**Comment by jordwalke** [comment]
generate.ml behaves the way it does because it compiles conditionals/branching by using `compile_argument_passing/parallel_renaming` which is used for closure compilation of variables, but I think it would require a slightly different version of those two functions for compiling conditionals/if/switches.

**Comment by hhugo** [comment]
Try `--enable vardecl`. I think I ve implemented exactly what ask for long time ago. I'm not sure it works fine though

**Comment by jordwalke** [comment]
Unfortunately, I seem to be hitting the `./compiler//lib/js_assign.ml:    failwith_ "Some variables escaped (#%d)" (S.cardinal (mapper#get_free))` case with that flag. It might be that some other flags are not playing well with it?

**Comment by jordwalke** [comment]
I couldn't find any tests covering that flag, so is it possible that feature broke? Also, I wonder if it would be better to implement that feature in `generate.ml` so that the code is in the most "normal" form possible - with variables declared at the top by default without having to enable a flag and perform another traversal on the tree. I also can see how much more difficult that would be because touching the logic that generates conditionals and switches seems a little daunting.

**Comment by hhugo** [comment]
As I said, it's possible that it doesn't work well - but it shouldn't be too hard to fix (adding tests at the same time).

Implementing this in `generate.ml` might be "better" but I'm not sure it's worth the cost.

**Comment by jordwalke** [comment]
I agree. I bet the fix is easy in `js_traverse.ml` since other traversals work fine. I think some of the variable names might not be getting copied over properly in the transform.

**Comment by jordwalke** [comment]
For posterity: If anyone ever does feel like implementing this in the conditional generation in `generate.ml` in the future, then it's also an opportunity to choose the zero value as the default for some of these variables which would simplify the output from:

```javascript
function extension_constructor(x) {
  if (is_block(x)) {
    if (caml_obj_tag(x) !== 248) {
      if (1 <= x.length - 1) {
        var slot = x[1];
        var switch__0 = 1;
      } else {
        var switch__0 = 0;
      }
    } else {
      var switch__0 = 0;
    }
  } else {
    var switch__0 = 0;
  }
  if (!switch__0) {
    var slot = x;
  }
}
```

into:

```javascript

Into:

```javascript
function extension_constructor(x) {
  if (is_block(x)) {
    if (caml_obj_tag(x) !== 248) {
      if (1 <= x.length - 1) {
        var slot = x[1];
        var switch__0 = 1;
      } else {
        var switch__0 = 0;
      }
    } else {
      var switch__0 = 0;
    }
  } else {
    var switch__0 = 0;
  }
  if (!switch__0) {
    var slot = x;
  }
}
```

Which is pretty trivially simplified by `js_simpl` into:
```javascript
function extension_constructor(x) {
  var switch__0 = 0;
  var slot = undefined;
  if (is_block(x) && caml_obj_tag(x) !== 248 && 1 <= x.length - 1) {
    slot = x[1];
    switch__0 = 1;
  }
  if (!switch__0) {
    slot = x;
  }
}
```

I agree it might not be worth the cost - just leaving the idea here for posterity.

**Comment by hhugo** [comment]
Would you mind trying #946 ?

**Comment by hhugo** [comment]
> For posterity: If anyone ever does feel like implementing this in the conditional generation in `generate.ml` in the future, then it's also an opportunity to choose the zero value as the default for some of these variables which would simplify the output from:
> 
> ```js
> function extension_constructor(x) {
>   if (is_block(x)) {
>     if (caml_obj_tag(x) !== 248) {
>       if (1 <= x.length - 1) {
>         var slot = x[1];
>         var switch__0 = 1;
>       } else {
>         var switch__0 = 0;
>       }
>     } else {
>       var switch__0 = 0;
>     }
>   } else {
>     var switch__0 = 0;
>   }
>   if (!switch__0) {
>     var slot = x;
>   }
> }
> ```
> 
> Into:
> 
> ```js
> function extension_constructor(x) {
>   var switch__0 = 0;
>   var slot = undefined;
>   if (is_block(x)) {
>     if (caml_obj_tag(x) !== 248) {
>       if (1 <= x.length - 1) {
>         slot = x[1];
>         switch__0 = 1;
>       } else {
>       }
>     } else {
>     }
>   } else {
>   }
>   if (!switch__0) {
>     slot = x;
>   }
> }
> ```
> 
> Which is pretty trivially simplified by `js_simpl` into:
> 
> ```js
> function extension_constructor(x) {
>   var switch__0 = 0;
>   var slot = undefined;
>   if (is_block(x) && caml_obj_tag(x) !== 248 && 1 <= x.length - 1) {
>     slot = x[1];
>     switch__0 = 1;
>   }
>   if (!switch__0) {
>     slot = x;
>   }
> }
> ```
> 
> I agree it might not be worth the cost - just leaving the idea here for posterity.

This has been implemented in #948 
Here is the generated code
```javascript
function extension_constructor(x) {
  var switch$0=0;
  if(is_block(x) && caml_obj_tag(x) !== 248 && 1 <= x.length - 1) {
    var slot=x[1];
    switch$0 = 1
  }
  if(! switch$0)
    var slot=x;
}
```

**Comment by jordwalke** [comment]
That looks like a huge improvement @hhugo.

I'm eager to test it and see if the resulting pattern in that output is now a common occurence:

```
function extension_constructor(x) {
  var switch$0=0;
  if(expr) {
    var slot=A;
    switch$0 = 1
  }
  if(!switch$0)
    var slot=B;
}
```

Because if so, then it wouldn't be difficult to implement if/else simplification that turns all of those occurrences into:

```
function extension_constructor(x) {
  var slot = (expr) ? A : B;
}
```
